crates/core/src/types.rs

```diff
@@ -324,10 +333,77 @@ impl TransactionWithStatusMeta {
+                                        })
+                                        .collect(),
+                                })
+                            }
+                        })
+                        .collect(),
+                ),
                 log_messages: Some(failure.meta.logs.clone()),
-                pre_token_balances: Some(vec![]),
-                post_token_balances: Some(vec![]),
+                pre_token_balances: {
+                    let balances: Vec<TransactionTokenBalance> = pre_token_accounts_with_indexes
+                        .iter()
+                        .zip(token_mints.clone())
+                        .zip(token_program_ids)
+                        .map(|(((i, a), mint), token_program)| TransactionTokenBalance {
+                            account_index: *i as u8,
+                            mint: a.mint().to_string(),
+                            ui_token_amount: UiTokenAmount {
+                                ui_amount: Some(format_ui_amount(a.amount(), mint.decimals())),
+                                decimals: mint.decimals(),
+                                amount: a.amount().to_string(),
+                                ui_amount_string: format_ui_amount_string(
+                                    a.amount(),
+                                    mint.decimals(),
+                                ),
+                            },
+                            owner: a.owner().to_string(),
+                            program_id: token_program.to_string(),
+                        })
+                        .collect();
+                    Some(balances.clone())
+                },
+                post_token_balances: {
+                    let balances: Vec<TransactionTokenBalance> = pre_token_accounts_with_indexes
+                        .iter()
+                        .zip(token_mints)
+                        .zip(token_program_ids)
+                        .map(|(((i, a), mint), token_program)| TransactionTokenBalance {
+                            account_index: *i as u8,
+                            mint: a.mint().to_string(),
+                            ui_token_amount: UiTokenAmount {
+                                ui_amount: Some(format_ui_amount(a.amount(), mint.decimals())),
+                                decimals: mint.decimals(),
+                                amount: a.amount().to_string(),
+                                ui_amount_string: format_ui_amount_string(
+                                    a.amount(),
+                                    mint.decimals(),
+                                ),
+                            },
+                            owner: a.owner().to_string(),
+                            program_id: token_program.to_string(),
+                        })
+                        .collect();
+                    Some(balances)
+                },
```

Comment:
Since pre/post will be the same here, can we just do this once at the top of the fn and clone?